crates/pera-core/src/dwallet_mpc/mpc_session.rs

```diff
@@ -0,0 +1,212 @@
+        party_id: PartyID,
+    ) -> DwalletMPCResult<(ConsensusTransaction, Vec<PartyID>)> {
+        let pending_messages = self.pending_messages.clone();
+        let (status, round) = match self.status {
+            MPCSessionStatus::Pending | MPCSessionStatus::FirstExecution => {
+                (MPCSessionStatus::Active(0), 0)
+            }
+            MPCSessionStatus::Active(round) => (MPCSessionStatus::Active(round + 1), round + 1),
+            _ => {
+                return Err(DwalletMPCError::MPCSessionError {
+                    session_id: self.session_info.session_id,
+                    error: format!(
+                        "failed to advance the MPC session, unexpected status: {}",
+                        self.status
+                    ),
+                })
+            }
+        };
+        self.status = status;
+        let advance_result = self.party.advance(
+            pending_messages,
+            self.session_info.flow_session_id,
+            party_id,
+            weighted_threshold_access,
+            self.public_input.clone(),
+        );
+
+        match advance_result {
+            Ok(AsynchronousRoundResult::Advance {
+                malicious_parties,
+                message,
+            }) => {
+                self.pending_messages.insert(round, HashMap::new());
+                Ok((
+                    self.new_dwallet_mpc_message(message).map_err(|e| {
+                        DwalletMPCError::MPCSessionError {
+                            session_id: self.session_info.session_id,
+                            error: format!(
+                                "failed to create a new MPC message on advance(): {:?}",
+                                e
+                            ),
+                        }
+                    })?,
+                    malicious_parties,
+                ))
+            }
+            Ok(AsynchronousRoundResult::Finalize {
+                malicious_parties,
+                private_output,
+                public_output,
+            }) => {
+                self.status = MPCSessionStatus::Finished(public_output.clone(), private_output);
+                Ok((
+                    self.new_dwallet_mpc_output_message(public_output)?,
+                    malicious_parties,
+                ))
+            }
+            Err(DwalletMPCError::MaliciousParties(malicious_parties)) => {
+                self.restart();
+                Err(DwalletMPCError::MaliciousParties(malicious_parties))
+            }
+            Err(e) => {
+                self.status = MPCSessionStatus::Failed;
+                Err(e)
+            }
+        }
+    }
+
+    /// A function to restart an MPC session.
+    /// Being called when session advancement has failed due to malicious parties.
+    /// Those parties will be flagged as malicious and ignored,
+    /// the session will be restarted.
+    fn restart(&mut self) {
+        self.status = MPCSessionStatus::FirstExecution;
```

Comment:
What is this? we are only supporting async protocols, where did you invent this behavior from? Looks like sync protocol behavior